btcrelay.se

```diff
@@ -0,0 +1,520 @@
+    bits = byte(0, wordWithBits) + byte(1, wordWithBits)*BYTES_1 + byte(2, wordWithBits)*BYTES_2 + byte(3, wordWithBits)*BYTES_3
+    target = targetFromBits(bits)
+
+    # we only check the target and do not do other validation (eg timestamp)
+    # to save gas
+    if blockHash > 0 && blockHash < target:
+        m_saveAncestors(blockHash, hashPrevBlock)  # increments ibIndex
+
+        save(self.block[blockHash]._blockHeader[0], blockHeaderBytes, chars=80)
+
+        blockHeight = m_getHeight(blockHash)
+        prevBits = m_getBits(hashPrevBlock)
+        if !m_difficultyShouldBeAdjusted(blockHeight) || self.ibIndex == 1:  # since blockHeight is 1 more than blockNumber; OR clause is special case for 1st header
+            # we need to check prevBits isn't 0 otherwise the 1st header
+            # will always be rejected (since prevBits doesn't exist for the initial parent)
+            # This allows blocks with arbitrary difficulty from being added to
+            # the initial parent, but as these forks will have lower score than
+            # the main chain, they will not have impact.
+            if bits != prevBits && prevBits != 0:
+                log(type=Failure, ERR_DIFFICULTY)
+                return(ERR_DIFFICULTY)
+        else:
+            prevTarget = targetFromBits(prevBits)
+            prevTime = m_getTimestamp(hashPrevBlock)
+
+            # (blockHeight - DIFFICULTY_ADJUSTMENT_INTERVAL) is same as [getHeight(hashPrevBlock) - (DIFFICULTY_ADJUSTMENT_INTERVAL - 1)]
+            startBlock = self.fastGetBlockHash(blockHeight - DIFFICULTY_ADJUSTMENT_INTERVAL)
+            startTime = m_getTimestamp(startBlock)
+
+            newBits = m_computeNewBits(prevTime, startTime, prevTarget)
+            if bits != newBits && newBits != 0:  # newBits != 0 to allow first header
+                log(type=Failure, ERR_RETARGET)
+                return(ERR_RETARGET)
+
+        difficulty = 0x00000000FFFF0000000000000000000000000000000000000000000000000000 / target # https://en.bitcoin.it/wiki/Difficulty
+        scoreBlock = scorePrevBlock + difficulty
+        m_setScore(blockHash, scoreBlock)
+
+        # equality allows block with same score to become an (alternate) Head, so that
+        # when an (existing) Head becomes stale, the chain can continue with the alternate Head
+        if scoreBlock >= self.highScore:
+            self.heaviestBlock = blockHash
+            self.highScore = scoreBlock
+
+        return(blockHeight)
+
+    return(0)
+
+
+# returns 1 if tx is in the block given by 'txBlockHash' and the block is
+# in Bitcoin's main chain (ie not a fork)
+#
+# the merkle proof is represented by 'txHash', 'txIndex', 'sibling', where:
+# - 'txHash' is the hash of the tx
+# - 'txIndex' is the index of the tx within the block
+# - 'sibling' are the merkle siblings of tx
+def verifyTx(txHash, txIndex, sibling:arr, txBlockHash):
+    if !self.feePaid(txBlockHash, m_getFeeAmount(txBlockHash), value=msg.value):  # in incentive.se
```

Comment:
Calling `feePaid` in `verifyTx` instead of `computeMerkle` allows an attacker to re-implement its own `verifyTx` without the fee check, therefore using the full functionality of btcrelay without ever paying a fee to block header relayers.